libraries/standard/mqtt/src/iot_mqtt_operation.c

```diff
@@ -301,6 +332,158 @@ static bool _scheduleNextRetry( _mqttOperation_t * pOperation )
 
 /*-----------------------------------------------------------*/
 
+static bool _sendKeepAlivePingReq( _mqttConnection_t * pMqttConnection )
+{
+    bool status = true;
+    uint32_t swapStatus = 0;
+    size_t bytesSent = 0;
+    _mqttOperation_t * pPingreqOperation = &( pMqttConnection->pingreq );
+
+    /* Swap status is not checked when asserts are disabled. */
+    ( void ) swapStatus;
+
+    /* Because PINGREQ may be used to keep the MQTT connection alive, it is
+     * more important than other operations. Bypass the queue of jobs for
+     * operations by directly sending the PINGREQ in this job. */
+    bytesSent = pMqttConnection->pNetworkInterface->send( pMqttConnection->pNetworkConnection,
+                                                          pPingreqOperation->u.operation.pMqttPacket,
+                                                          pPingreqOperation->u.operation.packetSize );
+
+    if( bytesSent != pPingreqOperation->u.operation.packetSize )
+    {
+        IotLogError( "(MQTT connection %p) Failed to send PINGREQ.", pMqttConnection );
+        status = false;
+    }
+    else
+    {
+        /* Assume the keep-alive will fail. The network receive callback will
+         * clear the failure flag upon receiving a PINGRESP. */
+        swapStatus = Atomic_CompareAndSwap_u32( &( pPingreqOperation->u.operation.periodic.ping.failure ),
+                                                1,
+                                                0 );
+        IotMqtt_Assert( swapStatus == 1 );
```

Comment:
This assert doesn't make sense to me. Isn't the purpose of this block to safely handle a known/expected race condition?

Reply:
I'm not entirely sure why the assert was part of the code if it's possible for the CAS to fail. Will update once I have more info.

Reply:
The swap here is expected to always succeed.  The logic around the `failure` flag is:

1. Create connection with `failure=0`
1. When a ping request is sent, set `failure=1` (what this line of code is doing)
1. Schedule a task to check for ping response some time in the future.
1. If a response is received, set `failure=0`.
1. Task for checking response runs. If it sees `failure=0`, this loop repeats.  If it sees `failure=1`, that means no response was received and the connection terminates.

The only reason this CAS can fail is if `failure=1` here.  That shouldn't be possible because the connection would have been terminated by the lack of ping response and this code should never run.  That's what the assert checks.

Reply:
Per our in-person discussion, it seems that swapStatus=0 could happen, however unlikely. While that's a pre-existing issue not directly related to this PR, there is a chance that this refactoring change could cause a latent bug of that nature to occur more frequently.